File: single_agent.py
import argparse
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import PillowWriter, FuncAnimation
import tqdm

class GridWorld:
    def __init__(self, size, n_mines, flag_pos):
        self.size = size
        self.n_mines = n_mines
        self.flag_pos = flag_pos
        self.grid = np.zeros((size, size))
        self.mines = self.place_mines()
        self.state = (0, 0)

    def place_mines(self):
        mines = []
        while len(mines) < self.n_mines:
            pos = (np.random.randint(0, self.size), np.random.randint(0, self.size))
            if pos not in mines and pos != (0, 0) and pos != self.flag_pos:
                mines.append(pos)
                self.grid[pos] = -1
        return mines

    def reset(self):
        self.state = (0, 0)
        return self.state

    def step(self, action):
        x, y = self.state
        if action == 0:  
            x = max(0, x - 1)
        elif action == 1:  
            x = min(self.size - 1, x + 1)
        elif action == 2:  
            y = max(0, y - 1)
        elif action == 3:  
            y = min(self.size - 1, y + 1)

        self.state = (x, y)

        if self.state in self.mines:
            return self.state, -10, True
        elif self.state == self.flag_pos:
            return self.state, 10, True
        else:
            return self.state, -1, False

class QLearningAgent:
    def __init__(self, env, alpha=0.1, gamma=0.9, epsilon=0.1):
        self.env = env
        self.alpha = alpha
        self.gamma = gamma
        self.epsilon = epsilon
        self.q_table = np.zeros((env.size, env.size, 4))

    def choose_action(self, state):
        if np.random.rand() < self.epsilon:
            return np.random.randint(4)
        x, y = state
        return np.argmax(self.q_table[x, y])

    def update(self, state, action, reward, next_state):
        x, y = state
        nx, ny = next_state
        best_next_action = np.max(self.q_table[nx, ny])
        self.q_table[x, y, action] += self.alpha * (
            reward + self.gamma * best_next_action - self.q_table[x, y, action]
        )

    def train(self, episodes):
        for _ in tqdm.tqdm(range(episodes), desc="Training"):
            state = self.env.reset()
            done = False
            while not done:
                action = self.choose_action(state)
                next_state, reward, done = self.env.step(action)
                self.update(state, action, reward, next_state)
                state = next_state

    def visualize_policy(self, output_gif, max_steps=100):
        if not output_gif:
            return
        fig, ax = plt.subplots(figsize=(20, 20))
        ax.set_xlim(-0.5, self.env.size - 0.5)
        ax.set_ylim(-0.5, self.env.size - 0.5)
        ax.set_xticks(range(self.env.size))
        ax.set_yticks(range(self.env.size))
        ax.grid()

        for mine in self.env.mines:
            ax.plot(mine[1], self.env.size - 1 - mine[0], 'rx', markersize=12, label="Mine")

        ax.plot(self.env.flag_pos[1], self.env.size - 1 - self.env.flag_pos[0], 'g*', markersize=15, label="Flag")
        agent_marker, = ax.plot([], [], 'bo', markersize=8, label="Agent")

        def init():
            agent_marker.set_data([], [])
            return agent_marker,

        def update(frame):
            x, y = frame
            agent_marker.set_data(y, self.env.size - 1 - x)
            return agent_marker,

        state = self.env.reset()
        frames = [state]
        done = False

        while not done and len(frames) < max_steps:
            action = np.argmax(self.q_table[state[0], state[1]])
            state, _, done = self.env.step(action)
            frames.append(state)

        ani = FuncAnimation(fig, update, frames=frames, init_func=init, blit=True, repeat=False)
        ani.save(output_gif, writer=PillowWriter(fps=2))

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--size", type=int, default=32)
    parser.add_argument("--n_mines", type=int, default=40)
    parser.add_argument("--flag_x", type=int, default=31)
    parser.add_argument("--flag_y", type=int, default=31)
    parser.add_argument("--episodes", type=int, default=20000)
    parser.add_argument("--alpha", type=float, default=0.1)
    parser.add_argument("--gamma", type=float, default=0.9)
    parser.add_argument("--epsilon", type=float, default=0.1)
    parser.add_argument("--gif_path", type=str, default="")
    parser.add_argument("--max_steps", type=int, default=100)
    args = parser.parse_args()

    env = GridWorld(args.size, args.n_mines, (args.flag_x, args.flag_y))
    agent = QLearningAgent(env, alpha=args.alpha, gamma=args.gamma, epsilon=args.epsilon)
    agent.train(args.episodes)
    agent.visualize_policy(args.gif_path, max_steps=args.max_steps)

if __name__ == "__main__":
    main()

File: single_agent.cpp
#include <algorithm>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <iostream>
#include <string>
#include <tuple>
#include <vector>

class GridWorld {
public:
  GridWorld(int size, int n_mines, std::pair<int, int> flag_pos)
      : size(size), n_mines(n_mines), flag_pos(flag_pos), state({0, 0}) {
    grid.resize(size, std::vector<int>(size, 0));
    std::srand((unsigned)std::time(nullptr));
    placeMines();
  }

  void placeMines() {
    mines.clear();
    while ((int)mines.size() < n_mines) {
      int x = std::rand() % size;
      int y = std::rand() % size;
      std::pair<int, int> pos = {x, y};
      if (std::find(mines.begin(), mines.end(), pos) == mines.end() &&
          (x != 0 || y != 0) && (x != flag_pos.first || y != flag_pos.second)) {
        mines.push_back(pos);
        grid[x][y] = -1;
      }
    }
  }

  std::pair<int, int> reset() {
    state = {0, 0};
    return state;
  }

  std::tuple<std::pair<int, int>, int, bool> step(int action) {
    int x = state.first;
    int y = state.second;
    if (action == 0) {
      x = (x > 0) ? x - 1 : x;
    } else if (action == 1) {
      x = (x < size - 1) ? x + 1 : x;
    } else if (action == 2) {
      y = (y > 0) ? y - 1 : y;
    } else if (action == 3) {
      y = (y < size - 1) ? y + 1 : y;
    }
    state = {x, y};
    for (auto &m : mines) {
      if (m.first == x && m.second == y) {
        return std::make_tuple(state, -10, true);
      }
    }
    if (x == flag_pos.first && y == flag_pos.second) {
      return std::make_tuple(state, 10, true);
    }
    return std::make_tuple(state, -1, false);
  }

  int getSize() const { return size; }

  bool isMine(int x, int y) const {
    for (auto &m : mines) {
      if (m.first == x && m.second == y)
        return true;
    }
    return false;
  }

  bool isFlag(int x, int y) const {
    return (x == flag_pos.first && y == flag_pos.second);
  }

private:
  int size;
  int n_mines;
  std::pair<int, int> flag_pos;
  std::vector<std::vector<int>> grid;
  std::vector<std::pair<int, int>> mines;
  std::pair<int, int> state;
};

class QLearningAgent {
public:
  QLearningAgent(GridWorld &env, double alpha = 0.1, double gamma = 0.9,
                 double epsilon = 0.1)
      : env(env), alpha(alpha), gamma(gamma), epsilon(epsilon) {
    int s = env.getSize();
    Q.resize(s,
             std::vector<std::vector<double>>(s, std::vector<double>(4, 0.0)));
    std::srand((unsigned)std::time(nullptr));
  }

  int chooseAction(const std::pair<int, int> &state) {
    double r = (double)std::rand() / RAND_MAX;
    if (r < epsilon) {
      return std::rand() % 4;
    }
    int x = state.first;
    int y = state.second;
    double bestValue = Q[x][y][0];
    int bestAction = 0;
    for (int a = 1; a < 4; a++) {
      if (Q[x][y][a] > bestValue) {
        bestValue = Q[x][y][a];
        bestAction = a;
      }
    }
    return bestAction;
  }

  void update(const std::pair<int, int> &state, int action, int reward,
              const std::pair<int, int> &next_state) {
    int x = state.first;
    int y = state.second;
    int nx = next_state.first;
    int ny = next_state.second;
    double bestNext = Q[nx][ny][0];
    for (int a = 1; a < 4; a++) {
      if (Q[nx][ny][a] > bestNext) {
        bestNext = Q[nx][ny][a];
      }
    }
    double tdTarget = reward + gamma * bestNext;
    Q[x][y][action] += alpha * (tdTarget - Q[x][y][action]);
  }

  void train(int episodes) {
    for (int ep = 0; ep < episodes; ep++) {
      auto state = env.reset();
      bool done = false;
      while (!done) {
        int action = chooseAction(state);
        auto stepResult = env.step(action);
        auto next_state = std::get<0>(stepResult);
        int reward = std::get<1>(stepResult);
        done = std::get<2>(stepResult);
        update(state, action, reward, next_state);
        state = next_state;
      }
    }
  }

  int bestAction(int x, int y) {
    double bestVal = Q[x][y][0];
    int bestAct = 0;
    for (int a = 1; a < 4; a++) {
      if (Q[x][y][a] > bestVal) {
        bestVal = Q[x][y][a];
        bestAct = a;
      }
    }
    return bestAct;
  }

  void printPolicy() {
    int s = env.getSize();
    for (int x = 0; x < s; x++) {
      for (int y = 0; y < s; y++) {
        if (env.isMine(x, y)) {
          std::cout << "M ";
          continue;
        }
        if (env.isFlag(x, y)) {
          std::cout << "F ";
          continue;
        }
        int act = bestAction(x, y);
        if (act == 0)
          std::cout << "^ ";
        else if (act == 1)
          std::cout << "v ";
        else if (act == 2)
          std::cout << "< ";
        else if (act == 3)
          std::cout << "> ";
      }
      std::cout << std::endl;
    }
  }

private:
  GridWorld &env;
  double alpha;
  double gamma;
  double epsilon;
  std::vector<std::vector<std::vector<double>>> Q;
};

int main(int argc, char **argv) {
  int size = 32;
  int n_mines = 40;
  int flag_x = 31;
  int flag_y = 31;
  int episodes = 20000;
  double alpha = 0.1;
  double gamma = 0.9;
  double epsilon = 0.1;

  for (int i = 1; i < argc; i++) {
    if (std::strcmp(argv[i], "--size") == 0 && i + 1 < argc) {
      size = std::atoi(argv[++i]);
    } else if (std::strcmp(argv[i], "--n_mines") == 0 && i + 1 < argc) {
      n_mines = std::atoi(argv[++i]);
    } else if (std::strcmp(argv[i], "--flag_x") == 0 && i + 1 < argc) {
      flag_x = std::atoi(argv[++i]);
    } else if (std::strcmp(argv[i], "--flag_y") == 0 && i + 1 < argc) {
      flag_y = std::atoi(argv[++i]);
    } else if (std::strcmp(argv[i], "--episodes") == 0 && i + 1 < argc) {
      episodes = std::atoi(argv[++i]);
    } else if (std::strcmp(argv[i], "--alpha") == 0 && i + 1 < argc) {
      alpha = std::atof(argv[++i]);
    } else if (std::strcmp(argv[i], "--gamma") == 0 && i + 1 < argc) {
      gamma = std::atof(argv[++i]);
    } else if (std::strcmp(argv[i], "--epsilon") == 0 && i + 1 < argc) {
      epsilon = std::atof(argv[++i]);
    }
  }

  GridWorld env(size, n_mines, {flag_x, flag_y});
  QLearningAgent agent(env, alpha, gamma, epsilon);
  agent.train(episodes);
  agent.printPolicy();

  return 0;
}
